File: Cpp/714.cpp
#include<iostream>
#include<vector>

using namespace std;

class Solution{
public:
	int maxProfit(vector<int>& prices,int fee){
		int s0=0,s1=INT_MIN;
		for(auto p:prices){
			int tmp=s0;
			s0=max(s0,s1+p);
			s1=max(s1,tmp-p-fee);
		}
		return s0;
	}
};	

int main(){
	vector<int> test={1,3,2,8,4,9};
	int fee=2;
	Solution s;
	cout<<s.maxProfit(test,fee)<<endl;

	return 0;
}


/*
s0 profit having no stock
s1 profit having one stock
fee=2
    stock    1   3  2  8  4  9
    index    0   1  2  3  4  5
    tmp  0   0   0  0  0  5  5
    s0   0   0   0  0  5  5  8
    s1   M  -3  -3 -3 -3  -1 -1

    =>s0 =8 the last moment we have no stock in hand.
*/

File: Python/714.py
# greedy

class Solution(object):
    def maxProfit(self, prices, fee):
        """
        :type prices: List[int]
        :type fee: int
        :rtype: int
        """
        n = len(prices)
        minP = prices[0]
        res = 0
        for i in range(1, n):
            if prices[i] < minP:
                minP = prices[i]
            elif prices[i] > minP + fee:
                res += prices[i] - minP - fee
                minP = prices[i] - fee
        return res


# dp
class Solution(object):
    def maxProfit(self, prices, fee):
        """
        :type prices: List[int]
        :type fee: int
        :rtype: int
        """
        n = len(prices)
        buy = [0]*n
        buy[0] = -prices[0]
        sell = [0]*n

        for i in range(1, n):
            sell[i] = max(sell[i-1], buy[i-1] + prices[i] -fee)
            buy[i] = max(sell[i-1] - prices[i], buy[i-1])
        # print sell
        # print buy
        return sell[-1]

